apps/vx-scan/backend/src/app.ts

```diff
@@ -35,14 +36,117 @@ const debug = rootDebug.extend('app');
 
 type NoParams = never;
 
+/**
+ * An interface for interacting with USB drives. We inject this into the app so
+ * that we can easily mock it in tests.
+ */
+export interface Usb {
+  getUsbDrives: () => Promise<UsbDrive[]>;
+}
+
+/**
+ * Possible errors that can occur during configuration (currently there's only one).
+ */
+export type ConfigurationError = 'no_ballot_package_on_usb_drive';
+
 function buildApi(
   machine: PrecinctScannerStateMachine,
   interpreter: PrecinctScannerInterpreter,
   workspace: Workspace,
+  usb: Usb,
   logger: Logger
 ) {
   const { store } = workspace;
+
   return grout.createApi({
+    async configureFromBallotPackageOnUsbDrive(): Promise<
+      Result<void, ConfigurationError>
+    > {
+      assert(!store.getElectionDefinition());
+      const [usbDrive] = await usb.getUsbDrives();
+      assert(usbDrive?.mountPoint !== undefined, 'No USB drive mounted');
+
+      const directoryPath = path.join(
+        usbDrive.mountPoint,
+        BALLOT_PACKAGE_FOLDER
+      );
+      if (!existsSync(directoryPath)) {
+        return err('no_ballot_package_on_usb_drive');
+      }
+
+      const files = await fs.readdir(directoryPath, { withFileTypes: true });
+      const ballotPackageFiles = files.filter(
+        (file) => file.isFile() && file.name.endsWith('.zip')
+      );
+      if (ballotPackageFiles.length === 0) {
+        return err('no_ballot_package_on_usb_drive');
+      }
+
+      const ballotPackageFilesWithStats = await Promise.all(
+        ballotPackageFiles.map(async (file) => {
+          const filePath = path.join(directoryPath, file.name);
+          return {
+            ...file,
+            filePath,
+            // Include file stats so we can sort by creation time
+            ...(await fs.lstat(filePath)),
+          };
+        })
+      );
+
+      const [mostRecentBallotPackageFile] = [
+        ...ballotPackageFilesWithStats,
+      ].sort((a, b) => b.ctime.getTime() - a.ctime.getTime());
+
+      const ballotPackage = await readBallotPackageFromBuffer(
+        await fs.readFile(mostRecentBallotPackageFile.filePath)
+      );
+
+      const ballotTemplatesWithImages = await Promise.all(
+        ballotPackage.ballots.map(async (ballotTemplate) => {
+          const layoutsWithImages: BallotPageLayoutWithImage[] = [];
+          for await (const { page, pageNumber } of pdfToImages(
+            ballotTemplate.pdf,
+            {
+              scale: 2,
+            }
+          )) {
+            const ballotPageLayout = find(
+              ballotTemplate.layout,
+              (l) => l.metadata.pageNumber === pageNumber
+            );
+            layoutsWithImages.push({ ballotPageLayout, imageData: page });
+          }
+          return { ...ballotTemplate, layoutsWithImages };
```

Comment:
This is potentially a lot of data to load into memory all at once. Our letter-size ballots are stored in the ballot package as 1224x1584 JPEGs. Loading them into memory using `canvas` takes about 7.4MB per page, or 14.8MB per ballot card. Some of our MS elections have up to 26 ballot styles with 1-card ballots, which is 14.8MB * 26 = 384.8MB. Better to process a single card at a time if you can manage it.

Reply:
Thinking about this more, this means we now have two places in the app where we're rendering the PDFs: one here and one in `Store`. It'd be great to change the API of `Store` such that it doesn't allow this partial loading of templates and we instead just reset them, add the PDFs and layout info to the db, then load from the db so we always use the same code path? Something like that, anyway.

Reply:
Thanks for pointing this out. The old code did this one at a time, so I think it makes sense to preserve that behavior given the memory implications.

As far as the changes to the store, that sounds fine to me. Do you think it's worth doing now, or is this all likely to change anyway? I know you had mentioned in the past that we might change how we do configuration

Reply:
Is this what you were imagining: 6eadf3690?

Reply:
I think it's okay to do the smaller change for now.

Reply:
Well, I already got the bigger change done haha

Reply:
Hah, thanks! The only critique I'd add is something that extends deep into existing code and isn't _that_ big a deal. But since you asked, I'd replace `setHmpbTemplates(ballotTemplates: BallotPackageEntry[])` with `setHmpbTemplates(ballotTemplates: Iterable<BallotPackageEntry>)`, then make the caller pass some sort of iterator so we don't have to load all the PDFs into memory at once. However, the existing code already does that and the PDFs are much smaller than the rendered images (O(100KB) each or about ~3MB for the 26 ballot style case), so it's probably fine.

I just wanted to bring this up because I think until recently we have been pretty unconcerned with memory and performance. We're starting to change that and I think it's worth pushing on things like this so instead of the default being "oh I am processing N of a thing so I need an array" maybe moving to "oh I am processing N of a thing so I need an iterable or a stream, or maybe an array if I know it'll always be fairly small".

Reply:
That makes sense. Agree it's a bit too big of a change to implement here right now, but I'm glad to get a better understanding of the general constraints. I'll add a note about memory perf to this code so we don't forget.